### ISSUE TITLE
jserve: a client that hangs up during prefill still holds the model, because peer_gone is only asked between tokens

### ISSUE THREAD

**Issue by xoloki** [xoloki/jlib#283]
#281 made a client that hangs up stop the generation: the per-token callback
asks `out.peer_gone()` and returns false. **Prefill produces no tokens**, so
during it that callback never runs and nothing is asked.

Found while measuring #195. A four-file jcode request -- about 15,000 real
tokens -- was abandoned by its client at 600 seconds, and the server was still
working on it afterwards:

```
server still healthy: 503      <- busy with a conversation nobody is listening to
```

The model stayed held for the rest of a prefill nobody would read, which is
exactly what #281 set out to prevent, one phase earlier.

## It is newly fixable, which is why it is worth filing now

Before #282 the prefill was one pass and there was nowhere to ask. Chunked
prefill gives it a natural boundary: between chunks, which for a 15,000-token
prompt is about sixty opportunities rather than none.

## What it needs

`ai::generate`'s only way back to its caller is `on_token`, and calling that
with something that is not a token would be a worse contract than the problem.
Two honest shapes:

- **A second callback** -- `std::function<bool()> still_wanted`, asked between
  chunks. Small, explicit, and says what it is for.
- **A cancel token** -- `sys::cancel_token`, which already exists and already
  means "end this", and which jserve could arm from the same place it learns
  the client went.

The second composes with what `sys` has; the first needs nothing new. Either
way jserve passes something that answers `peer_gone()`.

## Not urgent, and here is the honest reason

`try_acquire` (#280) means the server stays answering either way -- a held
model is a 503 rather than a hang. This costs a wasted prefill and a busy
model, not availability.

**Comment by xoloki** [comment]
## A second reason: it also bounds how long Ctrl-C takes

Came up reviewing juliet's fix for jserve calling `server::stop()` from a
signal handler. The shutdown path is:

```
stop() -> cancel_live() -> connection coroutine destroyed
       -> ~relay() -> join() -> stop() sets m_stop
       -> generation returns false from on_token -> thread ends
```

`relay::join()` sets the stop flag *before* joining, so a generation ends at
the next token boundary and shutdown costs one decode step -- 45-75 ms on
Qwen2.5-Coder 7B, measured end to end at about 100 ms.

**Except during prefill, where `on_token` is never called.** `generate.hh`
chunks a long prompt and runs `m.forward(part, ignored)` per chunk, discarding
logits; the callback loop starts after that. So the same gap this issue is
about -- `wanted()`/`peer_gone` only asked between tokens -- means a stop
landing during prefill waits for the entire prefill.

Measured on the 7B: `prefill512` is 1.69 s (q8_0) and 2.07 s (Q4_K_M).
Extrapolating by chunk count, not measured, a jcode-sized 4000-token prompt is
order 16 s of uninterruptible work.

So the fix is worth more than the held model it was filed for: asking the
predicate between prefill chunks bounds **both** the abandoned-client case and
shutdown latency, and the chunk boundary already exists since #285.
